# Agenda, week 5: Modules and packages

1. Q&A
2. What is a module?
    - Why do we want/need them?
    - What do they contain?
    - Using `import` to use a module's contents
3. The different forms of `import`
4. Developing our own module -- what's involved?
5. Python's standard library
    - What is it?
    - What does it contain?
    - How can we use it?
    - Library vs. builtins
6. Modules vs. packages
7. PyPI and pip
    - Downloading and installing packages from PyPI
    - Finding good vs. bad packages on PyPI
    - How do we use them?
8. What next?
    - Where can you learn more Python skills?
    - How can/should you learn more Python skills?
    - Practice, books, videos, strategies
    - And: Ask me anything!


# What is a module?

DRY -- the rule of "don't repeat yourself."

1. If you have several lines of code that are almost the same, then you can replace them with a loop.
2. If you have code that repeats itself in several places in your program, then you can replace them with a function.
3. If you have code that repeats itself in several different programs, then you can use a *library*.

Every programming language supports libraries. This way:

- If you have code that you'll use in several programs, you can write it once, and use it many times. (You can also debug it once, and have those fixes/improvements automatically get to be used in all programs at once.)
- You can write a library that other people will use, so that they don't have to reinvent the wheel.
- The reverse is true: You can use a library that other people have written, so that you don't have to reinvent the wheel.

In Python, we call our libraries *modules*. Modules provide us with all of these advantages.

(Sometimes, we call it a package. We'll talk about that later. Spoiler: A package is a directory/folder containing multiple module files.)

In addition to these advantages, Python modules also act as *namespaces*. The idea is that each module can define variables and functions with the same names, and they won't conflict with one another. A namespace is kind of like a last name for variables/functions. It means that we aren't going to have a "namespace collision."

It's a rare Python program that doesn't use any modules.

# What does a module contain?

Any Python code we want, but mostly/usually, it'll contain definitions:

- Variable definitions, for things we'll want to use frequently
- Function definitions, for functions we'll want to use frequently
- Class definitions, if you (or someone else) wants to define new types of data structures.

We load a module using the `import` statement, and that gives us access to those variables, functions, and classes.

# Example: The `random` module

If I want a random number, then I could write my own random-number generating function. But that's going to be very hard and time consuming, if I can even get it right.

Instead, I can use the `random` module that comes with Python (in the "standard library"), and I can use it.

If I want to load this module, I use the `import` statement.

In [1]:
import random

I just imported the `random` module!

- `import` is a reserved keyword in Python; you cannot use it as a variable or function name.
- It is not a function! It is a statement. That's why you don't need to use `()` around the module name.
- The module name that we put after `import` is *not* a string! It is not a filename! In many other programming languages, we load a module by stating the name of the file we want to load. In Python, we say what module variable we want to define, and Python figures out what filename we want from that.
- After this line, `random` is a variable that we have defined, and it contains a module value.
- Once we've imported the module, it is available to our program as much as we want.

But how do I use the module? 

A module is a container for other values -- variables, functions, and data types. Each of those names is available via the module by adding a `.` to the module name and then giving the name of the variable, function, or data type we want.

For example, there's a function called `randint` in the `random` module. To invoke it, we say

    random.randint(0, 100)    # this returns a random number from 0 through 100

    

In [4]:
random.randint(0, 100)

86

# How do I know what names are in a module?

A few possibilities:

1. If we're in Jupyter (or the equivalent), we can use the `dir` function on a module to get a list of names that are available on it.
    - names in `ALL CAPS` are "constants," meaning that you shouldn't change their values
    - names that `_start` with one `_` are considered "private," meaning that you shouldn't use them unless you are willing to risk things, because they could change
    - names that `__start_and_end__` with a double `__` are mostly for internal use, although you can use/set them if you know what you're doing
    - We still don't know what are functions vs. values vs. data types without investigating more.
2. Use the `help` function on our module to get its documentation.
3. Go to the Python documentation, at https://docs.python.org, and read about the module
4. If you're using a Python editor like PyCharm or VSCode, then hovering over the module name will let you get documentation for it, either inside of the editor or via a link to a Web site.

In [5]:
dir(random)

['BPF',
 'LOG4',
 'NV_MAGICCONST',
 'RECIP_BPF',
 'Random',
 'SG_MAGICCONST',
 'SystemRandom',
 'TWOPI',
 '_ONE',
 '_Sequence',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_accumulate',
 '_acos',
 '_bisect',
 '_ceil',
 '_cos',
 '_e',
 '_exp',
 '_fabs',
 '_floor',
 '_index',
 '_inst',
 '_isfinite',
 '_lgamma',
 '_log',
 '_log2',
 '_os',
 '_parse_args',
 '_pi',
 '_random',
 '_repeat',
 '_sha512',
 '_sin',
 '_sqrt',
 '_test',
 '_test_generator',
 '_urandom',
 'betavariate',
 'binomialvariate',
 'choice',
 'choices',
 'expovariate',
 'gammavariate',
 'gauss',
 'getrandbits',
 'getstate',
 'lognormvariate',
 'main',
 'normalvariate',
 'paretovariate',
 'randbytes',
 'randint',
 'random',
 'randrange',
 'sample',
 'seed',
 'setstate',
 'shuffle',
 'triangular',
 'uniform',
 'vonmisesvariate',
 'weibullvariate']

In [6]:
help(random)

Help on module random:

NAME
    random - Random variable generators.

MODULE REFERENCE
    https://docs.python.org/3.13/library/random.html

    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
        bytes
        -----
               uniform bytes (values between 0 and 255)

        integers
        --------
               uniform within range

        sequences
        ---------
               pick random element
               pick random sample
               pick weighted random sample
               generate random permutation

        distributions on the real line:
        ------------------------------
               uniform
               triangular
               normal (Gaussian)
               l

# Exercise: Guessing name

1. Use `random.randint` to choose a random integer between 0-100.
2. Ask the user, repeatedly, to guess the number.
    - If the user gets it, then congratulate them and stop the program. You can also tell them how many guesses it took. (That's module advanced.)
    - If the user guesses too high, then tell them, "too high!"
    - If the user guesses too low, then tell them, "too low!"
    - If the user enters a non-number, then scold them

Example:

    Guess: 50
    Too high!
    Guess: 25
    Too low!
    Guess: I hate this
    That is not a number; try again
    Guess: 32
    You got it in 3 tries!


In [10]:
import random

number = random.randint(0, 100)
guess_count = 0

while True:
    guess = input('Guess: ').strip()
    guess_count += 1

    if not guess.isdigit():
        print(f'{guess} is not numeric; try again')
        continue
        
    n = int(guess)

    if n == number:
        print('You got it!')
        break
    elif n < number:
        print('Too low!')
    else:
        print('Too high!')

print(f'You got it in {guess_count} guesses.')

Guess:  50


Too low!


Guess:  75


Too high!


Guess:  60


Too high!


Guess:  55


Too high!


Guess:  53


Too low!


Guess:  54


You got it!
You got it in 6 guesses.


In [9]:
# the help function can give us help on any function/method in Python

help(random.randint)  # we don't run the function with (), but we do pass it to help as an argument

Help on method randint in module random:

randint(a, b) method of random.Random instance
    Return random integer in range [a, b], including both end points.



# More about our module -- the object itself and how Python loaded it

When we use `import`, we're assigning a module value to our variable (in this case, `random`). We can inspect that variable, asking it to print itself.

In [11]:
random

<module 'random' from '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/random.py'>

# How did Python know to look there?

If we say `import random`, then Python takes the name of the module (`random`), adds `.py` to it, and then looks for a file called `random.py`. But where does it look?

- First in the current directory -- here, that's wherever Jupyter is running, but it's generally wherever you run the program from.
- Then it looks through each directory in `sys.path`. Yes, that's a value in the `sys` module. To see it, you need to `import sys`.

In [12]:
import sys

sys.path    # it's a list of strings, with each string being a file/directory where we should look for modules

['/Users/reuven/.pyenv/versions/3.13.2/lib/python313.zip',
 '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13',
 '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/lib-dynload',
 '',
 '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/site-packages']

In [ ]:
# CC
# I have made a different start in my program is wrong? 

import random

secret = random.randint(0,10)
guess_count = 0
guess = -1

while guess != secret:
    guess = input(f"Indovina il numero:")
    guess_count += 1
    if not guess.isdigit():
        print(f"Sorry, {guess} is not a digit!")
        continue
        
    guess = int(guess)
    
    if guess > secret:
        print("To high!")
    elif guess < secret:
        print("To low!")
    else:
        print(f"Congratulation the secret number is {guess}\nYou made {guess_count} attempts")

In [13]:
random

<module 'random' from '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/random.py'>

In [16]:
# this is where the file was loaded from

random.__file__

'/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/random.py'

In [17]:
import os

os.path.split(random.__file__)  # os.path.split takes a string, and returns a 2-element tuple with the directory and filename

('/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13', 'random.py')

# What happens when you `import` a module?

1. Python checks to see if the module was already imported. If so, then it skips to step 3.
2. Python looks for the module in each of the directories in `sys.path`. If it finds a matching filename, then it loads that file as a module. For example, if it finds `random.py`, then it loads that file's contents.
3. Python assigns the module's contents to our module variable.

# Variations on `import`

- `import MODULE` -- this is the easiest, and most common, way to `import`.
- `from MODULE import NAME` -- this loads the entire module into memory, but doesn't define `MODULE` as a variable. Rather, it defines `NAME` instead, as a top-level variable rather than attribute of the module.
- `import MODULE as ALIAS` -- import the module, but define the variable to be `ALIAS` instead.
- `from MODULE import NAME as ALIAS` -- import the module, only define one name from the module, but use an alias instead
- `from MODULE import *` -- this asks Python to load the module, and then define every variable/function from the module in the current namespace as a global variable/function. **THIS IS AN ABSOLUTELY TERRIBLE THING TO DO!** 

In [18]:
# I've got random.randint
# what if I'm going to use it a lot? It'll be annoying to say random.randint each time.
# can I just say randint?

randint(0, 100)   # randint isn't a function at the top level -- it's within the "random" module

NameError: name 'randint' is not defined

In [19]:
# I can ask Python to define not the module as a variable, but our function instead

from random import randint

# in the above code:
# (1) we're still loading the entire "random" module into memory
# (2) the only variable that is defined is "randint" -- the "random" module variable is *not* defined in this syntax
# (3) this syntax exists to make it easier to refer to commonly called functions

In [20]:
randint(0, 100)

87

In [21]:
# Another variation on import: we can change the name of the variable that's defined

import random as r   # this means: find and load random.py, but don't define "random" as a variable; use "r" instead

# (1) maybe there's another variable/function named random, and you want to avoid a potential class
# (2) maybe it's easier to read/debug/work with using another name
# (3) sometimes, if you have a very deep hierarchy, it's easier to give a name rather than use a.b.c.d.e.f.g
# (4) everyone else does it

Example: Everyone in the data-analysis world uses NumPy and Pandas with aliases:



In [22]:
import numpy as np
import pandas as pd

In [23]:
import random as r

In [24]:
r.randint(0, 100)

4

# Next up

- We'll write our own (simple) module
- We'll see more of what happens when we `import` a module

# Creating a module

- We have to create a file
- The file's name must be a legitmate Python variable name before the `.` and `py` after the `.`.
- The file contains legal Python, particularly/usually variable and function definitions

In Jupyter I can actually create a file using a special, Web-based editor!

In [25]:
# let's try to load mymod.py, which is in the same directory

import mymod

In [26]:
mymod

<module 'mymod' from '/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/mymod.py'>

In [27]:
# before we add any functionality to the module, let's take a look at what names (attributes) it defines

dir(mymod)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__']

In [28]:
mymod.__file__

'/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/mymod.py'

In [30]:
mymod.__name__

'mymod'

In [31]:
import mymod

In [32]:
dir(mymod)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__']

In [33]:
# if we have already loaded a module, and we want to reload it, we need to use
# the importlib module, which comes with Python

# you should never be using reload in a real-life Python program!
# but... if you're using Jupyter and exploring, then it's very useful

from importlib import reload    # this function reloads a module, even if it was loaded before

reload(mymod) 

<module 'mymod' from '/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/mymod.py'>

In [34]:
dir(mymod)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'hello',
 'x',
 'y']

In [35]:
mymod.x

5

In [36]:
mymod.y

[10, 20, 30]

In [37]:
mymod.hello('world')

'Hello, world, from mymod!'

# Exercise: `menu` module

1. We often need to ask users to choose from among several different options in a program. We're going to write a function `menu` that will be inside of a module `menu` (yes, that'll make it `menu.menu`) that will take a list of strings -- the options that a user might be able to enter.
2. When we call the function, it'll ask the user to choose from the options in the list.
3. If the user chooses one of the legal options, then the function returns that value.
4. If the user chooses an illegal option, then the function scolds them and has them try again.
5. Create this `menu` module, which will contain only our function definition. Then, from outside of the module, `import` it and invoke the function, and show that it works.

In [38]:
import menu

s = menu.menu(['a', 'b', 'c'])

print(f'You chose {s}')

Choose from ['a', 'b', 'c']:  x


x is invalid; try again


Choose from ['a', 'b', 'c']:  C


C is invalid; try again


Choose from ['a', 'b', 'c']:  abc


abc is invalid; try again


Choose from ['a', 'b', 'c']:  b


You chose b


# A few more things

I've modified `mymod.py` such that it has two `print` calls, one at the start of the module and one at its end. Will this print anything when I `import` (or in our case, `reload`) the module?

In [39]:
reload(mymod)

Hello from mymod!
Done loading mymod!


<module 'mymod' from '/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/mymod.py'>

In addition: We've seen that any variable we define in our module (e.g., `x`, `y`, and `hello`) are all available to us outside of the module as attributes on the module -- e.g., `mymod.x`, `mymod.y`, and `mymod.hello`.

- Inside of the module, they're just variables
- Outside of the module, they are attributes on the module name

This raises the question: What about the attributes on the module name that Python defined, such as `__name__`? Are they available inside of the module as variables?

In [40]:
dir(mymod)

['__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'hello',
 'x',
 'y']

In [41]:
# now let's reload the module. What will it print for us?

reload(mymod)

Hello from mymod!
Done loading mymod!


<module 'mymod' from '/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/mymod.py'>

What do we see from this?

That Python defines a number of special variables inside of a module that we can use to report on and inspect our module from the inside, inlcuding `__name__`, which has the name of the module in it.

But it gets better!

# `__name__` is special

The variable `__name__` is always defined in Python. It can have one of two types of values:

- If it's used in a file that someone imported (i.e., a module), then it has the module's name as a string.
- If it's in the first file that was executed by Python (i.e., if the program was run directly), then it has the special string `'__main__'`, meaning that it's the "main" namespace of the program.

When we said `import mymod`, `__name__` was the string `'mymod'`. But when we ran `mymod.py` as a Python program, then it was the first thing loaded, and thus `__name__` had the value `'__main__'`.

Who cares?

Consider:

- When we `import` a module, the module is executed, from start to finish.
- We can include any Python code we want.
- If the module was imported, `__name__` will have a different value than if the program was executed.

This leads us to one of the most famous lines in all of Python:

```python
if __name__ == '__main__':
```

This code is at the bottom of nearly every module in Python. It basically says: Only run the below code if the module was run as a program. Do *not* run the code below this line if the module was imported.

This allows us to write a module that is both loadable/usable by others for its definitions, and also a standalone program that we can run and use.

In [42]:
reload(mymod)

<module 'mymod' from '/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/mymod.py'>

# Who cares?

Many, *many* modules take advantage of this:

- They can provide a demo of the module's functionality when the module is run as a program
- They can run a function, getting input from the user
- They can run automated tests on themselves when invoked

Many people believe that you must have this line in a Python program. You don't! It's a great convenience, but it's far from mandatory. 

# Don't confuse `__name__` and `'__main__'`!

- `__name__` is a variable. It contains a string value. Python sets this variable based on how you invoke a program.
- `'__main__'` is a string. It's the string value that is set to `__name__` when you're in the first program loaded by Python.

Both of these are called "dunder-SOMETHING," where "dunder" is Python slang for "double underscore, before and after."

# Exercise: Make `menu.py` runnable on its own

Modify `menu.py`, such that if we invoke it from the command line (i.e., not via `import`), then it will ask the user to enter one of `a`, `b`, or `c`, and will print the result.

In [43]:
reload(menu)

<module 'menu' from '/Users/reuven/Courses/Current/OReilly-2025-02Feb-python-5-steps/menu.py'>

# Next up

- Python's standard library
- Modules vs. packages
- PyPI and pip
- Where to next? (And AMA)

# Python standard library

This is a collection of modules that comes with Python. No matter what platform you're running on, you are guaranteed to have the same Python standard library available to you.

The standard library, because it comes with Python, is standardized for each version of Python. So if you are using Python 3.13, and your users are using Python 3.13, then you'll have the same standard library available -- which means that you can use `import` in your program and know what will be available to import.

Fair warning: The standard library is *very* big. Getting a good gut sense of what's in it will help.

The standard library is documented at https://docs.python.org.

In [44]:
# to use a module from the standard library, we just have to import it. Then we have access to the variables,
# functions, and data types defined in that module.

import os

os.listdir('/etc/')   # this will return a list of files in /etc/

['xinetd.d-migrated2launchd',
 'ssh_config.system_default',
 'ssh_config.applesaved',
 'periodic',
 'manpaths',
 'services~previous',
 'rc.common',
 'csh.logout~orig',
 'auto_master',
 'php.ini.default-5.2-previous~orig',
 'csh.login',
 'syslog.conf',
 'rtadvd.conf~previous',
 'syslog.conf~previous',
 'krb5.keytab',
 'sudoers.d',
 'bash_completion.d',
 'ssl',
 'kern_loader.conf.applesaved',
 'ttys~previous',
 'csh.logout',
 'aliases.db',
 'hosts.lpd',
 'bashrc_Apple_Terminal',
 'racoon',
 'snmp',
 'zshrc_Apple_Terminal',
 'named.conf.applesaved',
 'gettytab',
 'master.passwd~orig',
 'kern_loader.conf',
 'authorization.user_modified',
 'networks~orig',
 'paths.d',
 'asl',
 'csh.login~orig',
 'rtadvd.conf',
 'security',
 'protocols~previous',
 'group',
 'printcap',
 'auto_home',
 'php.ini.default-previous',
 'sudoers~',
 'manpaths.d',
 'smb.conf.applesaved',
 'ppp',
 'shells',
 'pear.conf-previous',
 'crontab',
 'slpsa.conf.applesaved',
 'rc.common~previous',
 'xinetd.d',
 'ttys',
 'grou

In [46]:
# If I'm using the terminal, I can say that I want all of the *.conf files
# why can't I do that in Python?

# I can, using another module, the glob module, which hanldes such patterns

import glob
glob.glob('/etc/*.conf')

['/etc/syslog.conf',
 '/etc/kern_loader.conf',
 '/etc/rtadvd.conf',
 '/etc/pf.conf',
 '/etc/launchd.conf',
 '/etc/autofs.conf',
 '/etc/slpsa.conf',
 '/etc/ntp_opendirectory.conf',
 '/etc/resolv.conf',
 '/etc/nfs.conf',
 '/etc/asl.conf',
 '/etc/ntp.conf',
 '/etc/AFP.conf',
 '/etc/man.conf',
 '/etc/newsyslog.conf',
 '/etc/notify.conf']

# Exercise: Globbing files

1. Ask the user to enter a pattern for filenames. This will be a string.
2. Iterate over each of the filenames that match the pattern the user gave, using `glob.glob`.
3. Find the size of that file, and print it next to the filename. (Note: One easy way to get the length is to iterate over each line, adding it to the total size for that file.)

Strategy:
- Use a `for` loop for each filename
- Use a `for` loop for each line in the opened file

Example:

    Enter pattern: /etc/*.conf
    /etc/syslog.conf: 1000
    /etc/kern_loader.conf 2000

If you enter names of non-text files, the program will crash, so try to make your patterns tight enough that they just match some text files.   

In [49]:
import glob

pattern = input('Enter filename pattern: ').strip()

for one_filename in glob.glob(pattern):
    f = open(one_filename)

    length = 0

    for one_line in f:
        length += len(one_line)

    print(f'{one_filename}: {length}')

Enter filename pattern:  /etc/[a-z]*.conf


/etc/syslog.conf: 96
/etc/kern_loader.conf: 0
/etc/rtadvd.conf: 891
/etc/pf.conf: 1027
/etc/launchd.conf: 19
/etc/autofs.conf: 1932
/etc/slpsa.conf: 52
/etc/ntp_opendirectory.conf: 23
/etc/resolv.conf: 368
/etc/nfs.conf: 43
/etc/asl.conf: 1051
/etc/ntp.conf: 27
/etc/man.conf: 2451
/etc/newsyslog.conf: 1318
/etc/notify.conf: 557


# What's wrong with the standard library?

1. It's only released/updated when there is a new Python release.
2. It's managed by the core Python development team. This means that it can't be too big.
3. It's released along with Python, so it definitely cannot be too big.

As a result, we have a second place where you can get Python modules. Those are actually "packages," entire directories that contain one or more modules, along with some metadata.

PyPI, the Python Package Index, is where we get these things. PyPi.org is a great resource for the Python world!

In [50]:
import rich

In [51]:
rich

<module 'rich' from '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/site-packages/rich/__init__.py'>

In [52]:
sys.path

['/Users/reuven/.pyenv/versions/3.13.2/lib/python313.zip',
 '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13',
 '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/lib-dynload',
 '',
 '/Users/reuven/.pyenv/versions/3.13.2/lib/python3.13/site-packages']

In [53]:
rich.print('Hello to everyone out there!')

Hello to everyone out there!

In [54]:
rich.print('[green]Hello[/green] to [yellow on purple]everyone[/yellow on purple] [italic]out there[/italic]!')

Hello to everyone out there!

If `pip` doesn't work in your terminal, try 

    python3 -m pip install THING

That should help `pip` to run.

# Things to consider about PyPI

1. Anyone can contribute! So there is a lot of bad code.
2. Finding the best packages can be challening. (We'll talk about this in a bit.)
3. Some packages might be security issues!
    - How popular is a project?
    - How long has it been around?
    - How many users does it have?
    - How many contributors does it have?
    - Check GitHub stars, commits, and the number of developers who have contributed
4. What license is used for the project?
    - Everything on PyPI is open source... but there are different open-source licenses
    - GNU Public License tends to come under fire more -- it is "viral," and if used in another project, makes the whole project open source, such that anyone can copy, modify, redistribute, or sell it.

# Next up

- Practice a bit with PyPI and `pip`
- A little bit about `uv`, a new alternative to `pip`
- Awesome Python -- how to evaluate packages
- Where to from here? And ... AMA

# Installing with `pip`

`pip` is a program that comes with Python, and you normally run it from the command line as

    pip install THING

or 

    python3 -m pip install THING

If you're inside of Jupyter already, then you can use a "magic command" starting with `%` to run `pip` from inside of Jupyter.

    %pip install THING

After you do one of the above, you should be able to say

    import THING

# Exercise: Using `rich`

1. Use `pip` to install `rich`.
2. Use `rich.print` to display some text
    - Include some `[color]` and `[/color]` instructions inside of the text you give to `rich.print`
    - Include an emoji by having `:name:` inside of the string you give to `rich.print`.

In [55]:
%pip install -U rich

Note: you may need to restart the kernel to use updated packages.


In [58]:
import rich

rich.print(f'Hello [red]this[/red] [green]is[/green] [purple]fun![/purple], :smile:')

Hello this is fun!, 😄

# What next?

The most important thing to do is: Practice as much as you can.

- Do projects (even small ones!) that are interesting/useful for you.
- Challenge yourself!

Also take classes, read books, etc., but doing things is the best way to learn.

If you want to do testing, you should learn "pytest". 

A few other things:
- My site, https://PracticeYourPython.com, has a list of where you can get exercises to improve (includes my books and online courses)
- Go to conferences!
    - PyCon US is in Pittsburgh in mid-May
    - Euro Python is in Prague in mid-July
    - There are lots of small, regional conferences and meetups, and they are generally VERY friendly and helpful and encouraging
- PyLadies -- encouraging of women in getting to coding with Python (and related technologies)



# My stuff 

1. I have three newsletters I publish each week:
    - Better developers, about Python (https://BetterDevelopersWeekly.com)
    - Bamboo Weekly, with challenges in data analysis using Python + Pandas (https://BambooWeekly.com)
    - Trainer Weekly, about being a trainer in software (https://TrainerWeekly.com)
2. I can come to your company
3. You can check out my learning platform at https://LernerPython.com 

Amazing UV lecture by Charlie Marsh: https://www.youtube.com/watch?v=gSKTfG1GXYQ